src/ui/src/builder/BuilderFieldsTools.vue

```diff
@@ -0,0 +1,282 @@
+				></BuilderEmbeddedCodeEditor>
+			</template>
+			<template v-if="toolForm.type == 'graph'">
+				<WdsTextInput v-model="toolForm.graphId"></WdsTextInput>
+			</template>
+			<div>
+				<WdsButton variant="builder" @click="saveToolForm"
+					>Save</WdsButton
+				>
+			</div>
+		</BuilderModal>
+	</div>
+</template>
+
+<script setup lang="ts">
+import { toRefs, inject, computed, ref, Ref, ComputedRef } from "vue";
+import { Component, FieldControl } from "@/writerTypes";
+import { useComponentActions } from "./useComponentActions";
+import injectionKeys from "../injectionKeys";
+import WdsButton from "@/wds/WdsButton.vue";
+import BuilderModal, { ModalAction } from "./BuilderModal.vue";
+import WdsTextareaInput from "@/wds/WdsTextareaInput.vue";
+import WdsTextInput from "@/wds/WdsTextInput.vue";
+import WdsDropdownInput from "@/wds/WdsDropdownInput.vue";
+import BuilderEmbeddedCodeEditor from "./BuilderEmbeddedCodeEditor.vue";
+
+type FunctionTool = {
+	type: "function";
+	description: string;
+	parameters: Record<
+		string,
+		{
+			type: string;
+			description: string;
+		}
+	>;
+};
+
+type GraphTool = {
+	type: "graph";
+	graph_ids: string[];
+};
+
+type Tool = FunctionTool | GraphTool;
+
+type ToolForm = {
+	isShown: boolean;
+	type: "function" | "graph";
+	name: string;
+	code: string;
+	graphId: string;
+};
+
+const wf = inject(injectionKeys.core);
+const ssbm = inject(injectionKeys.builderManager);
+const { setContentValue } = useComponentActions(wf, ssbm);
+
+const initFunctionToolCode = `
+{
+	"description": "An example tool",
+	"parameters": {
+		"city": string
+	}
+}
+`.trim();
+
+const toolFormInitValue = {
+	isShown: false,
+	type: "function" as "function" | "graph",
+	name: "new_tool",
+	code: initFunctionToolCode,
+	graphId: "999-999",
+};
+
+const toolForm: Ref<ToolForm> = ref(toolFormInitValue);
+
+const props = defineProps<{
+	componentId: Component["id"];
+	fieldKey: string;
+}>();
+const { componentId, fieldKey } = toRefs(props);
+const component = computed(() => wf.getComponentById(componentId.value));
+
+const tools: ComputedRef<Record<string, Tool>> = computed(() => {
+	let value = {};
+	try {
+		value = JSON.parse(component.value.content[fieldKey.value]);
+	} catch {
+		value = {};
+	}
+	return value;
+});
+
+function resetAndShowToolFormModal() {
+	toolForm.value = {
+		...toolFormInitValue,
+		isShown: true,
+	};
+}
+
+function getToolFromForm(): Tool {
+	const { type, code, graphId } = toolForm.value;
+	if (type == "function") {
+		return {
+			...JSON.parse(code),
+			type,
+		};
+	}
+	if (type == "graph") {
+		return {
+			type,
+			graph_ids: [graphId],
+		};
+	}
+	throw "Unexpected tool type.";
+}
+
+function validateToolForm(form: ToolForm): string[] {
+	let errors = [];
+	const { name } = form;
+	if (!name) {
+		errors.push("The name cannot be empty.");
+	}
+	if (Object.keys(tools.value).includes(name)) {
+		errors.push("An existing tool with the specified name already exists.");
+	}
+	return errors;
+}
```

Comment:
Optional, but I think it's a better pattern to use `computed` here. For example.
```suggestion
const tooolFromForm = computed<Tool | null>(() => {
	const { type, code, graphId } = toolForm.value;
	if (type == "function") {
		return {
			...JSON.parse(code),
			type,
		};
	}
	if (type == "graph") {
		return {
			type,
			graph_ids: [graphId],
		};
	}
	return null;
});

const toolFormErrors = computed<string[]>(() => {
	let errors = [];
	const { name } = form;
	if (!name) {
		errors.push("The name cannot be empty.");
	}
	if (Object.keys(tools.value).includes(name)) {
		errors.push("An existing tool with the specified name already exists.");
	}
	return errors;
});
```